source/blog/2016-04-01-deadlock-investigation.html.md

```diff
@@ -0,0 +1,191 @@
+
+```diff
+diff --git a/actionpack/lib/action_dispatch/middleware/executor.rb b/actionpack/lib/action_dispatch/middleware/executor.rb
+index 06245b4..4f24fe1 100644
+--- a/actionpack/lib/action_dispatch/middleware/executor.rb
++++ b/actionpack/lib/action_dispatch/middleware/executor.rb
+@@ -7,12 +7,30 @@ def initialize(app, executor)
+     end
+
+     def call(env)
++Rails.logger.info "XXX #{Thread.current.inspect} #{__method__} before run!"
++begin
+       state = @executor.run!
++rescue => err
++  Rails.logger.info "XXX #{Thread.current.inspect} #{__method__} run! rescue, before app.call, error: #{err}."
++  raise
++ensure
++  Rails.logger.info "XXX #{Thread.current.inspect} #{__method__} run! ensure, before app.call"
++end
++
++Rails.logger.info "XXX #{Thread.current.inspect} #{__method__} after run!, before app.call"
+       begin
+         response = @app.call(env)
++Rails.logger.info "XXX #{Thread.current.inspect} #{__method__} after app.call, before state.complete!"
+         returned = response << ::Rack::BodyProxy.new(response.pop) { state.complete! }
++Rails.logger.info "XXX #{Thread.current.inspect} #{__method__} after state.complete!, returned: #{returned.inspect}"
++returned
+       ensure
+-        state.complete! unless returned
++        unless returned
++Rails.logger.info "XXX #{Thread.current.inspect} #{__method__} before ensure state.complete!, returned: #{returned.inspect}"
++          ret = state.complete!
++Rails.logger.info "XXX #{Thread.current.inspect} #{__method__} after ensure state.complete!, returned: #{returned.inspect}"
++          ret
++        end
+       end
+     end
+   end
+```
+
+Here is a snippet:
+
+```
+[----] I, [2016-03-31T18:08:16.029485 #7579:3e35e24]  INFO -- : Processing by ApiController#show as JSON
+[----] I, [2016-03-31T18:08:16.029638 #7579:3e35e24]  INFO -- :   Parameters: {"filter"=>["retired=false"], "collection"=>"services"}
+[----] I, [2016-03-31T18:08:16.054986 #7579:3e35c08]  INFO -- : Processing by ApiController#show as JSON
+[----] I, [2016-03-31T18:08:16.055079 #7579:3e35c08]  INFO -- :   Parameters: {"filter"=>["retired=true"], "collection"=>"services"}
+[----] I, [2016-03-31T18:08:25.938834 #7579:3d7460c]  INFO -- : XXX #<Thread:0x00000007ae8c18@/opt/rubies/ruby-2.2.4/lib/ruby/gems/2.2.0/gems/puma-2.16.0/lib/puma/thread_pool.rb:61 run> call run! rescue, before app.call, error: could not obtain a connection from the pool within 10.000 seconds (waited 10.001 seconds); all pooled connections were in use.
+```
+
+The important part is this:
+```
+...call run! rescue, before app.call, error: could not obtain a connection from the pool within 10.000 seconds
+```
+
+So, the key takeaways:
+ * The rescue around the `@executor.run!` caught and logged a connection pool timeout exception before re-raising
+ * This is exactly what Matthew guessed and will require a rework of the interlock to ensure exceptions raised while we have this interlock ensure the lock is released
```

Comment:
As a "how to debug" narrative, it might be better framed with a mention of the four possible scenarios I outlined, which our debugging was designed to distinguish. I did correctly guess which one seemed most likely, but our process wasn't quite "step 2, correctly guess exactly what the problem is" :wink: